test/inference.jl

```diff
@@ -1101,4 +1101,51 @@ let isa_tfunc = Core.Inference.t_ffunc_val[
         @test isa_tfunc(c, Type{Complex}) === Const(false)
         @test isa_tfunc(c, Type{Complex{T}} where T) === Const(false)
     end
+    @test isa_tfunc(Val{1}, Type{Val{T}} where T) === Bool
+    @test isa_tfunc(Val{1}, DataType) === Bool
+    @test isa_tfunc(Any, Const(Any)) === Const(true)
+    @test isa_tfunc(Any, Union{}) === Const(false)
+    @test isa_tfunc(Any, Type{Union{}}) === Bool
+    @test isa_tfunc(Union{Int64, Float64}, Type{Real}) === Const(true)
+    @test isa_tfunc(Union{Int64, Float64}, Type{Integer}) === Bool
+    @test isa_tfunc(Union{Int64, Float64}, Type{AbstractArray}) === Const(false)
+end
+
+let subtype_tfunc = Core.Inference.t_ffunc_val[
+        findfirst(Core.Inference.t_ffunc_key, <:)][3]
+    @test subtype_tfunc(Type{<:Array}, Const(AbstractArray)) === Const(true)
+    @test subtype_tfunc(Type{<:Array}, Type{AbstractArray}) === Const(true)
+    @test subtype_tfunc(Type{<:Array}, Type{AbstractArray{Int}}) == Bool
+    @test subtype_tfunc(Type{<:Array{Real}}, Type{AbstractArray{Int}}) === Const(false)
+    @test subtype_tfunc(Type{<:Array{Real, 2}}, Const(AbstractArray{Real, 2})) === Const(true)
+    @test subtype_tfunc(Type{Array{Real, 2}}, Const(AbstractArray{Int, 2})) === Const(false)
+    @test subtype_tfunc(DataType, Int) === Bool
+    @test subtype_tfunc(DataType, Const(Type{Int})) === Bool
+    @test subtype_tfunc(DataType, Const(Type{Array})) === Bool
+    @test subtype_tfunc(UnionAll, Const(Type{Int})) === Bool
+    @test subtype_tfunc(UnionAll, Const(Type{Array})) === Bool
+    @test subtype_tfunc(Union, Const(Union{Float32, Float64})) === Bool
+    @test subtype_tfunc(Union, Type{Union}) === Bool
+    @test subtype_tfunc(Union{}, Const(Int)) === Const(true) # any result is ok
+    @test subtype_tfunc(Union{}, Const(Union{})) === Const(true) # any result is ok
+    @test subtype_tfunc(Union{}, typeof(Union{})) === Bool # any result is ok
+    @test subtype_tfunc(Union{}, Union{}) === Const(true) # any result is ok
```

Comment:
I wouldn't expect this to be called on `Union{}` much, if ever. Probably should test `Type{Union{}}` instead of or in addition to this.